### PR TITLE
Logger

### PR DESCRIPTION
**Pull request by xtda**
# Description

Increases the flexibility of the built logging capabilities this is part of work to improve error handling and move to a "fail and recover" instead "fail and fatal" architecture but the changes were starting to get large and make code review a nightmare and I wanted to get the logging out

This replaces the Standard Print/Fatal/Panic with additional logging types:

**DEBUG:** 

For all debugging information that is good to now but not critical such as a process completing 
An example of this is when verbose is enabled request information is displayed as a debug

Debug("Debug")
Debugf("Debug formatted string %s, somestring)
Debugln("Debug this will be new line terminated")

**INFO:**

For all information events at the moment this is only used for showing orderbook and ticket outputs

Info("Info")
Infof("Info formatted string %s", somestring)
Infoln("Info this will be new line terminated")

**WARN:**

For any issues that wont cause a failure in program execution or flow 

An example of this is warning a user that websocket support is disabled for an exchange
Good to know but not critical 

Warn("Warn")
Warnf("Warn %s", somestring)

**ERROR:**

These are errors that will break expected flow but can be recovered from 

Error("Error")
Error("Error %s", somestring)

Examples of this are a temporary connection timeout or once off unexpected response 
 
**FATAL:**

Fatal("Fatal followed by a os.Exit(1)")
Fatal("Fatal followed by a os.Exit(1) %s", somestring)

These are errors that will break program flow and **cannot** not be recovered from further execution must be halted

Warn/Error/Fatal include no ln option to follow practice to not new line terminate errors

All of these levels can be turned on or off via the level configuration option separated via a |

There is also Print/Printf/Println that are are just aliased to the default logger this is good to use when you want thread-safe console output but don't need any additional formatting/log to file (such as prompting user for a password)

On platforms that support proper ANSI colouring you can enable coloured output to allow you to easily spot different log types
image

if enabled you can also output to files this will save in the gocryptodatadir/logs folder under the name
you can also enable log rotation at the moment this is primitive and simply renames the log if enabled and detected on start up to the current date - time 

An example configuration is:

 "logging": {
  "enabled": true,
  "file": "debug.txt",
  "colour": true,
  "level": "DEBUG|INFO|WARN|ERROR|FATAL",
  "rotate": true
 }

Fixes # (issue)

## Type of change

Please delete options that are not relevant and add an `x` in `[]` as item is complete.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

I am flagging this as a _breaking change_ because it does technically change existing logging functionality 

I have replaced/updated all existing usage of the default logger with this outside of the tools 
  
# How Has This Been Tested?

Automated tests have been written to test correct flow/handling of errors these ca be rerun via 
go test ./logger/...

Manual testing of configuration options and possible failures also happened on Ubuntu 18.04 & Windows 10 Machines running Go 1.14.1

Benchmarking has also been Included

## Please also consider improving test coverage whilst working on a certain package

# Checklist:
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool 
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally and on Travis with my changes
- [x] Any dependent changes have been merged and published in downstream modules

- [x] My code follows the style guidelines of this project

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -959,6 +960,44 @@ func (c *Config) RetrieveConfigCurrencyPairs(enabledOnly bool) error {
...
 
+// CheckLoggerConfig checks to see logger values are present and valid in config
+// if not creates a default instance of the logger
+func (c *Config) CheckLoggerConfig() (err error) {
+	m.Lock()
+	defer m.Unlock()
+
+	// check if enabled is nil or level is a blank string
+	if c.Logging.Enabled == nil || c.Logging.Level == "" {
+		// Creates a new pointer to bool and sets it as false
```
Comment (shazbert) [inline config/config.go]: Following ln sets bool pointer as true

File: exchanges/alphapoint/alphapoint.go [hunk 2]
```diff
@@ -191,7 +191,7 @@ func (a *Alphapoint) CreateAccount(firstName, lastName, email, phone, password s
 
 	err := a.SendAuthenticatedHTTPRequest("POST", alphapointCreateAccount, request, &response)
 	if err != nil {
-		log.Println(err)
+		log.Error(err)
```
Comment (shazbert) [inline exchanges/alphapoint/alphapoint.go]: Might be better to return that error

File: exchanges/alphapoint/alphapoint_websocket.go [hunk 3]
```diff
@@ -20,25 +20,25 @@ func (a *Alphapoint) WebsocketClient() {
 		a.WebsocketConn, _, err = Dialer.Dial(a.WebsocketURL, http.Header{})
```
Comment (shazbert) [inline exchanges/alphapoint/alphapoint_websocket.go]: rm ln, my bad.

File: exchanges/alphapoint/alphapoint_websocket.go [hunk 4]
```diff
@@ -20,25 +20,25 @@ func (a *Alphapoint) WebsocketClient() {
...
+			log.Errorf("%s Unable to connect to Websocket. Error: %s\n", a.Name, err)
 			continue
 		}
 
 		if a.Verbose {
-			log.Printf("%s Connected to Websocket.\n", a.Name)
+			log.Debugf("%s Connected to Websocket.\n", a.Name)
 		}
 
 		err = a.WebsocketConn.WriteMessage(websocket.TextMessage, []byte(`{"messageType": "logon"}`))
```
Comment (shazbert) [inline exchanges/alphapoint/alphapoint_websocket.go]: rm ln also my bad.

File: exchanges/coinbasepro/coinbasepro.go [hunk 5]
```diff
@@ -14,9 +13,11 @@ import (
 	"github.com/thrasher-/gocryptotrader/common"
 	"github.com/thrasher-/gocryptotrader/config"
 	"github.com/thrasher-/gocryptotrader/currency/symbol"
-	"github.com/thrasher-/gocryptotrader/exchanges"
+	exchange "github.com/thrasher-/gocryptotrader/exchanges"
 	"github.com/thrasher-/gocryptotrader/exchanges/request"
 	"github.com/thrasher-/gocryptotrader/exchanges/ticker"
+
```
Comment (shazbert) [inline exchanges/coinbasepro/coinbasepro.go]: rm ln

File: exchanges/coinut/coinut_websocket.go [hunk 6]
```diff
@@ -3,16 +3,17 @@ package coinut
...
 	"net/url"
 	"time"
 
 	"github.com/gorilla/websocket"
 	"github.com/thrasher-/gocryptotrader/common"
 	"github.com/thrasher-/gocryptotrader/currency/pair"
-	"github.com/thrasher-/gocryptotrader/exchanges"
+	exchange "github.com/thrasher-/gocryptotrader/exchanges"
 	"github.com/thrasher-/gocryptotrader/exchanges/orderbook"
+
```
Comment (shazbert) [inline exchanges/coinut/coinut_websocket.go]: rm ln

File: exchanges/coinut/coinut_wrapper.go [hunk 7]
```diff
@@ -3,17 +3,17 @@ package coinut
...
-	"github.com/thrasher-/gocryptotrader/currency/symbol"
-
 	"github.com/thrasher-/gocryptotrader/common"
 	"github.com/thrasher-/gocryptotrader/currency/pair"
-	"github.com/thrasher-/gocryptotrader/exchanges"
+	"github.com/thrasher-/gocryptotrader/currency/symbol"
+	exchange "github.com/thrasher-/gocryptotrader/exchanges"
 	"github.com/thrasher-/gocryptotrader/exchanges/orderbook"
 	"github.com/thrasher-/gocryptotrader/exchanges/ticker"
+
```
Comment (shazbert) [inline exchanges/coinut/coinut_wrapper.go]: rm ln

File: exchanges/gateio/gateio_wrapper.go [hunk 8]
```diff
@@ -2,15 +2,16 @@ package gateio
...
 	"strconv"
 	"sync"
 
 	"github.com/thrasher-/gocryptotrader/common"
 	"github.com/thrasher-/gocryptotrader/currency/pair"
-	"github.com/thrasher-/gocryptotrader/exchanges"
+	exchange "github.com/thrasher-/gocryptotrader/exchanges"
 	"github.com/thrasher-/gocryptotrader/exchanges/orderbook"
 	"github.com/thrasher-/gocryptotrader/exchanges/ticker"
+
```
Comment (shazbert) [inline exchanges/gateio/gateio_wrapper.go]: rm ln

File: exchanges/hitbtc/hitbtc.go [hunk 9]
```diff
@@ -13,9 +12,11 @@ import (
 	"github.com/thrasher-/gocryptotrader/common"
 	"github.com/thrasher-/gocryptotrader/config"
 	"github.com/thrasher-/gocryptotrader/currency/symbol"
-	"github.com/thrasher-/gocryptotrader/exchanges"
+	exchange "github.com/thrasher-/gocryptotrader/exchanges"
 	"github.com/thrasher-/gocryptotrader/exchanges/request"
 	"github.com/thrasher-/gocryptotrader/exchanges/ticker"
+
```
Comment (shazbert) [inline exchanges/hitbtc/hitbtc.go]: rm ln

**Comment by gloriousCode** [comment]
Thanks xtda, This is a great update for log monitoring, especially when using services. Thanks for analysing all of our logs and updating their level appropriately.

File: exchanges/huobi/huobi_websocket.go [hunk 10]
```diff
@@ -15,8 +14,10 @@ import (
 	"github.com/gorilla/websocket"
 	"github.com/thrasher-/gocryptotrader/common"
 	"github.com/thrasher-/gocryptotrader/currency/pair"
-	"github.com/thrasher-/gocryptotrader/exchanges"
+	exchange "github.com/thrasher-/gocryptotrader/exchanges"
 	"github.com/thrasher-/gocryptotrader/exchanges/orderbook"
+
```
Comment (shazbert) [inline exchanges/huobi/huobi_websocket.go]: rm ln

File: exchanges/huobi/huobi_wrapper.go [hunk 11]
```diff
@@ -3,16 +3,17 @@ package huobi
...
 	"sync"
 
 	"github.com/thrasher-/gocryptotrader/common"
 	"github.com/thrasher-/gocryptotrader/config"
 	"github.com/thrasher-/gocryptotrader/currency/pair"
-	"github.com/thrasher-/gocryptotrader/exchanges"
+	exchange "github.com/thrasher-/gocryptotrader/exchanges"
 	"github.com/thrasher-/gocryptotrader/exchanges/orderbook"
 	"github.com/thrasher-/gocryptotrader/exchanges/ticker"
+
```
Comment (shazbert) [inline exchanges/huobi/huobi_wrapper.go]: rm ln

File: exchanges/huobihadax/huobihadax_wrapper.go [hunk 12]
```diff
@@ -3,15 +3,17 @@ package huobihadax
...
 	"strconv"
 	"sync"
 
 	"github.com/thrasher-/gocryptotrader/common"
 	"github.com/thrasher-/gocryptotrader/currency/pair"
-	"github.com/thrasher-/gocryptotrader/exchanges"
+	exchange "github.com/thrasher-/gocryptotrader/exchanges"
 	"github.com/thrasher-/gocryptotrader/exchanges/orderbook"
 	"github.com/thrasher-/gocryptotrader/exchanges/ticker"
+
```
Comment (shazbert) [inline exchanges/huobihadax/huobihadax_wrapper.go]: rm ln

File: exchanges/okcoin/okcoin.go [hunk 13]
```diff
@@ -811,7 +811,7 @@ func (o *OKCoin) GetFuturesUserInfo() {
 	err := o.SendAuthenticatedHTTPRequest(okcoinFuturesUserInfo, url.Values{}, nil)
```
Comment (shazbert) [inline exchanges/okcoin/okcoin.go]: rm ln

File: exchanges/okcoin/okcoin.go [hunk 14]
```diff
@@ -823,7 +823,7 @@ func (o *OKCoin) GetFuturesPosition(symbol, contractType string) {
 	err := o.SendAuthenticatedHTTPRequest(okcoinFuturesPosition, v, nil)
```
Comment (shazbert) [inline exchanges/okcoin/okcoin.go]: rm ln

File: exchanges/okcoin/okcoin.go [hunk 15]
```diff
@@ -841,7 +841,7 @@ func (o *OKCoin) FuturesTrade(amount, price float64, matchPrice, leverage int64,
 	err := o.SendAuthenticatedHTTPRequest(okcoinFuturesTrade, v, nil)
```
Comment (shazbert) [inline exchanges/okcoin/okcoin.go]: rm ln

File: exchanges/okcoin/okcoin.go [hunk 16]
```diff
@@ -856,7 +856,7 @@ func (o *OKCoin) FuturesBatchTrade(orderData, symbol, contractType string, lever
 	err := o.SendAuthenticatedHTTPRequest(okcoinFuturesTradeBatch, v, nil)
```
Comment (shazbert) [inline exchanges/okcoin/okcoin.go]: rm ln

File: exchanges/okcoin/okcoin.go [hunk 17]
```diff
@@ -870,7 +870,7 @@ func (o *OKCoin) CancelFuturesOrder(orderID int64, symbol, contractType string)
 	err := o.SendAuthenticatedHTTPRequest(okcoinFuturesCancel, v, nil)
```
Comment (shazbert) [inline exchanges/okcoin/okcoin.go]: rm ln

File: exchanges/okex/okex_wrapper.go [hunk 18]
```diff
@@ -3,7 +3,7 @@ package okex
 import (
 	"errors"
 	"fmt"
-	"log"
+
```
Comment (shazbert) [inline exchanges/okex/okex_wrapper.go]: rm ln

File: exchanges/poloniex/poloniex.go [hunk 19]
```diff
@@ -168,7 +168,7 @@ func (p *Poloniex) GetOrderbook(currencyPair string, depth int) (OrderbookAll, e
 			return oba, err
 		}
 		if len(resp.Error) != 0 {
-			log.Println(resp.Error)
+			log.Error(resp.Error)
```
Comment (shazbert) [inline exchanges/poloniex/poloniex.go]: I dont think we need this at all because we are returning it.

File: logger/logger.go [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+		}
+	} else {
+		clearAllLoggers()
+	}
+	return
+}
+
+// setDefaultOutputs() this setups defaults used by the logger
+// This allows it to be used without any user configuration
+
```
Comment (shazbert) [inline logger/logger.go]: rm ln

File: logger/logger.go [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+		log.Ldate|log.Ltime)
+
+	fatalLogger = log.New(os.Stdout,
+		"[FATAL]: ",
+		log.Ldate|log.Ltime)
+}
+
+// colorOutput() sets the prefix of each log type to matching colour
+// TODO: add windows support
+
```
Comment (shazbert) [inline logger/logger.go]: rm ln

File: logger/logger.go [hunk 22]
```diff
@@ -0,0 +1,124 @@
...
+		debugLogger.SetPrefix("\033[34m[DEBUG]\033[0m: ")
+		infoLogger.SetPrefix("\033[32m[INFO]\033[0m: ")
+		warnLogger.SetPrefix("\033[33m[WARN]\033[0m: ")
+		errorLogger.SetPrefix("\033[31m[ERROR]\033[0m: ")
+		fatalLogger.SetPrefix("\033[31m[FATAL]\033[0m: ")
+	}
+}
+
+// clearAllLoggers() sets all logger flags to 0 and outputs to Discard
+
```
Comment (shazbert) [inline logger/logger.go]: rm ln

File: logger/logger.go [hunk 23]
```diff
@@ -0,0 +1,124 @@
...
+	debugLogger.SetOutput(ioutil.Discard)
+	infoLogger.SetOutput(ioutil.Discard)
+	warnLogger.SetOutput(ioutil.Discard)
+	errorLogger.SetOutput(ioutil.Discard)
+	fatalLogger.SetOutput(ioutil.Discard)
+}
+
+// setupOutputs() sets up the io.writer to use for logging
+// TODO: Fix up rotating at the moment its a quick job
+
```
Comment (shazbert) [inline logger/logger.go]: rm ln

File: logger/loggers.go [hunk 24]
```diff
@@ -0,0 +1,80 @@
...
+
+// Infoln handler takes any input infoLogger returns formatted output to infoLogger writer
+func Infoln(v ...interface{}) {
+	infoLogger.Println(v...)
+}
+
+// Print aliased to Standard log.Print
+var Print = log.Print
+
+// Printf alaised to tandard log.Printf
```
Comment (shazbert) [inline logger/loggers.go]: misspelling aliased and standard

File: logger/loggers.go [hunk 25]
```diff
@@ -0,0 +1,80 @@
...
+var Printf = log.Printf
+
+// Println  alaised to tandard log.Println
```
Comment (shazbert) [inline logger/loggers.go]: misspelling

File: logger/loggers.go [hunk 26]
```diff
@@ -0,0 +1,80 @@
...
+func Error(v ...interface{}) {
+	errorLogger.Print(v...)
+}
+
+// Errorf handler takes any input returns unformatted output to errorLogger writer
+func Errorf(data string, v ...interface{}) {
+	errorLogger.Printf(data, v...)
+}
+
+// Fatal  handler takes any input returns unformatted output to fatalLogger writer
```
Comment (shazbert) [inline logger/loggers.go]: rm space between fatal and handler

File: logger/loggers.go [hunk 27]
```diff
@@ -0,0 +1,80 @@
...
+func Fatal(v ...interface{}) {
+	// Send to Output instead of Fatal to allow us to increase the output depth by 1 to make sure the correct file is displayed
+	fatalLogger.Output(2, fmt.Sprint(v...))
+	os.Exit(1)
+}
+
+// Fatalf  handler takes any input returns unformatted output to fatalLogger writer
```
Comment (shazbert) [inline logger/loggers.go]: rm extra space between fatalf and handler

File: config/config.go (hunk 1 above)
Comment (xtda) [reply config/config.go]: Fixed thanks

File: exchanges/coinbasepro/coinbasepro.go (hunk 5 above)
Comment (xtda) [reply exchanges/coinbasepro/coinbasepro.go]: Fixed thanks

**Comment by shazbert** [comment]
tACK windows 10 test

**Pull request merged**